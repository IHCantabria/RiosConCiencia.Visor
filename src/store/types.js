//DATAHUB

//Mutations
export const M_SET_PRODUCTS = "datahub/M_SET_PRODUCTS";
export const M_SET_SOURCES = "datahub/M_SET_SOURCES";
export const M_SET_PRODUCTMOMENTS = "datahub/M_SET_PRODUCTMOMENTS";
export const M_SET_PRODUCTVARIABLETYPES = "datahub/M_SET_PRODUCTVARIABLETYPES";
export const M_UPDATE_PRODUCTS_SELECTED = "datahub/M_UPDATE_PRODUCTS_SELECTED";

//Actions
export const A_FETCH_DATAHUB_PRODUCTS = "datahub/A_FETCH_DATAHUB_PRODUCTS";
export const A_FETCH_DATAHUB_SOURCES = "datahub/A_FETCH_DATAHUB_SOURCES";
export const A_FETCH_DATAHUB_PRODUCTMOMENTS =
  "datahub/M_SET_DATAHUB_PRODUCTMOMENTS";
export const A_FETCH_DATAHUB_PRODUCTVARIABLETYPES =
  "datahub/A_FETCH_DATAHUB_PRODUCTVARIABLETYPES";
export const A_SET_CLEAR_PRODUCTS = "datahub/A_SET_CLEAR_PRODUCTS";
export const A_UPDATE_PRODUCTS_SELECTED = "datahub/A_UPDATE_PRODUCTS_SELECTED";

//Getters
export const G_GET_PRODUCTS_SORTED = "datahub/G_GET_PRODUCTS_SORTED";
export const G_GET_PRODUCTS_SELECTED = "datahub/G_GET_PRODUCTS_SELECTED";
export const G_GET_PRODUCTS_COUNT = "datahub/G_GET_PRODUCTS_COUNT";
export const G_GET_PRODUCT_BY_MAPID = "datahub/G_GET_PRODUCT_BY_MAPID";
export const G_GET_SOURCE_NAMESHORT = "datahub/G_GET_SOURCE_NAMESHORT";
export const G_GET_SOURCE_TYPE = "datahub/G_GET_SOURCE_TYPE";
export const G_GET_PRODUCTMOMENTS_SORTED =
  "datahub/G_GET_PRODUCTMOMENTS_SORTED";
export const G_GET_PRODUCTVARIABLETYPES_BY_PRODUCTMOMENTS =
  "datahub/G_GET_PRODUCTVARIABLETYPES_BY_PRODUCTMOMENTS";

//APISENSORS

//Mutations
export const M_SET_SENSORS = "apiSensors/M_SET_SENSORS";
export const M_SET_BUOYS = "apiSensors/M_SET_BUOYS";
export const M_SET_STATIONS = "apiSensors/M_SET_STATIONS";
export const M_SET_BUOYTYPES = "apiSensors/M_SET_BUOYTYPES";
export const M_SET_STATIONTYPES = "apiSensors/M_SET_STATIONTYPES";

//Actions
export const A_FETCH_APISENSORS_BUOYS = "apiSensors/A_FETCH_APISENSORS_BUOYS";
export const A_FETCH_APISENSORS_STATIONS =
  "apiSensors/A_FETCH_APISENSORS_STATIONS";
export const A_FETCH_APISENSORS_BUOYTYPES =
  "apiSensors/A_FETCH_APISENSORS_BUOYTYPES";
export const A_FETCH_APISENSORS_STATIONTYPES =
  "apiSensors/A_FETCH_APISENSORS_STATIONTYPES";
export const A_SET_SENSORS = "apiSensors/A_SET_SENSORS";
export const A_SET_CLEAR_SENSORS = "apiSensors/A_SET_CLEAR_SENSORS";

//Getters
export const G_GET_SENSORS_SORTED = "apiSensors/G_GET_SENSORS_SORTED";
export const G_GET_SENSORS_SELECTED = "apiSensors/G_GET_SENSORS_SELECTED";
export const G_GET_SENSORS_COUNT = "apiSensors/G_GET_SENSORS_COUNT";
export const G_GET_BUOYTYPE_NAME = "apiSensors/G_GET_BUOYTYPE_NAME";

//COORDS

//Mutations
export const M_SET_SEARCH_COORDS = "coords/M_SET_SEARCH_COORDS";
export const M_SET_GO_TO_COORDS = "coords/M_SET_GO_TO_COORDS";

//Actions
export const A_SET_SEARCH_COORDS = "coords/A_SET_SEARCH_COORDS";
export const A_SET_GO_TO_COORDS = "coords/A_SET_GO_TO_COORDS";

//FILTERS

//Mutations
export const M_SET_FILTER_MOMENT = "filters/M_SET_FILTER_MOMENT";
export const M_SET_FILTER_VARIABLE = "filters/M_SET_FILTER_VARIABLE";
export const M_SET_FILTER_BUOYSTYPE = "filters/M_SET_FILTER_BUOYSTYPE";
export const M_SET_FILTER_STATIONSTYPE = "filters/M_SET_FILTER_STATIONSTYPE";
export const M_SET_FILTER_ONLY_SPATIAL_FILTERS =
  "filters/M_SET_FILTER_ONLY_SPATIAL_FILTERS";
export const M_SET_FILTER_ALLOW_PRODUCTS =
  "filters/M_SET_FILTER_ALLOW_PRODUCTS";
export const M_SET_FILTER_ALLOW_BUOYS = "filters/M_SET_FILTER_ALLOW_BUOYS";
export const M_SET_FILTER_ALLOW_STATIONS =
  "filters/M_SET_FILTER_ALLOW_STATIONS";

//Actions
export const A_SET_FILTER_MOMENT = "filters/A_SET_FILTER_MOMENT";
export const A_SET_FILTER_VARIABLE = "filters/A_SET_FILTER_VARIABLE";
export const A_SET_FILTER_BUOYSTYPE = "filters/A_SET_FILTER_BUOYSTYPE";
export const A_SET_FILTER_STATIONSTYPE = "filters/A_SET_FILTER_STATIONSTYPE";
export const A_SET_FILTER_ONLY_SPATIAL_FILTERS =
  "filters/A_SET_FILTER_ONLY_SPATIAL_FILTERS";
export const A_SET_FILTER_ALLOW_PRODUCTS = "filters/_SET_FILTER_ALLOW_PRODUCTS";
export const A_SET_FILTER_ALLOW_BUOYS = "filters/A_SET_FILTER_ALLOW_BUOYS";
export const A_SET_FILTER_ALLOW_STATIONS =
  "filters/A_SET_FILTER_ALLOW_STATIONS";

//GETTERS
export const G_GET_FILTERS_READY = "filters/G_GET_FILTERS_READY";
