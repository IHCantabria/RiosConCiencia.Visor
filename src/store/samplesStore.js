import { defineStore } from "pinia";
import {
  getYears,
  getCampaignFilter,
  getDownloadFilters,
} from "@/utils/data-operator-utils.js";
import { CAMPAIGN_VALUES } from "@/config/data-config";
import { ADMIN_ROLE_ID } from "@/config/role-config.js";
import { shallowRef } from "vue";

export const useSamplesStore = defineStore("samplesStore", {
  state: () => {
    return {
      samples: null,
      samplesPicts: null,
      yearOptions: [],
      yearFilter: {},
      campaignOptions: [],
      campaignFilter: {},
      downloadOptions: [],
      downloadFilters: [],
      user: null,
      riverSectionHistoricData: {
        active: false,
        riverSectionId: null,
      },
      riversGeoJson: shallowRef(null),
    };
  },
  getters: {
    getSamplesFiltered() {
      return this.samples
        ? this.samples.filter(
            (sample) =>
              sample.year == this.yearFilter.name &&
              sample.idCampaign == this.campaignFilter.id,
          )
        : [];
    },
    getSamplesFilteredDownloadIds() {
      if (!this.samples) {
        return [];
      } else if (
        !this.downloadFilters.length ||
        (this.downloadFilters.length == 1 && this.downloadFilters[0].id == 0)
      ) {
        return this.samples.map((sample) => sample.id);
      } else {
        return this.samples
          .filter((sample) => this.getFiltersIds.includes(sample.idFilter))
          .map((sample) => sample.id);
      }
    },
    getSamplesPictsFilteredDownloadIds() {
      if (!this.samplesPicts) {
        return [];
      } else if (
        !this.downloadFilters.length ||
        (this.downloadFilters.length == 1 && this.downloadFilters[0].id == 0)
      ) {
        return this.samplesPicts.map((sample) => sample.id);
      } else {
        return this.samplesPicts
          .filter((sample) => this.getFiltersIds.includes(sample.idFilter))
          .map((sample) => sample.id);
      }
    },
    getFiltersIds() {
      return this.downloadFilters.length
        ? this.downloadFilters.map((filter) => filter.id)
        : [];
    },
    getIsFilterAll() {
      return this.downloadFilters[0].id == 0;
    },
    getUserToken() {
      return this.user ? this.user.token : "";
    },
    getUserLogged() {
      return this.user;
    },
    getAdminLogged() {
      return this.user && this.user.roleId == ADMIN_ROLE_ID;
    },
    getSamplesLoaded() {
      return this.samples;
    },
  },
  actions: {
    setSamples(operation) {
      if (!operation) return;
      this.samples = operation;
      const years = getYears(operation);
      const filterCampaign = getCampaignFilter(
        operation,
        years[0],
        CAMPAIGN_VALUES,
      );
      this.yearOptions = years;
      this.yearFilter = years[0];
      this.campaignOptions = CAMPAIGN_VALUES;
      this.campaignFilter = filterCampaign;
      const downloadOptions = getDownloadFilters(years, CAMPAIGN_VALUES);
      this.downloadOptions = downloadOptions;
      this.downloadFilters = [downloadOptions[0]];
    },
    setPictsSamples(operation) {
      this.samplesPicts = operation;
    },
    setRiversGeoJson(operation) {
      this.riversGeoJson = operation;
    },
    setUser(operation) {
      this.user = operation;
    },
    setSampleCampaignFilter(operation) {
      this.campaignFilter = operation;
    },
    setSampleYearFilter(operation) {
      this.yearFilter = operation;
    },
    setSampleDownloadFilters(operation) {
      this.downloadFilters = operation;
    },
    async fetchSamplesAdmin(operation) {
      this.setSamplesAdmin(operation);
    },
    async fetchSamples(operation) {
      this.setSamples(operation);
    },
    async fetchPictsSamples(operation) {
      this.setPictsSamples(operation);
    },
    async fetchUser(operation) {
      this.setUser(operation);
    },
    async setSampleCampaignFilterAction(operation) {
      this.setSampleCampaignFilter(operation);
    },
    async setSampleYearFilterAction(operation) {
      this.setSampleYearFilter(operation);
    },
    async setSampleDownloadFiltersAction(operation) {
      this.setSampleDownloadFilters(operation);
    },
    setRiverSectionHistoricData(idRiverSection) {
      if (!idRiverSection) {
        this.riverSectionHistoricData.active = false;
        this.riverSectionHistoricData.riverSectionId = null;
        return;
      }
      this.riverSectionHistoricData.active = true;
      this.riverSectionHistoricData.riverSectionId = idRiverSection;
    },
  },
});
