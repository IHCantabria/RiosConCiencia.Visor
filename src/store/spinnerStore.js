import { defineStore } from "pinia";
import { spinnerConfig, spinnerEnums } from "@/config/spinner-config";

export const useSpinnerStore = defineStore("spinnerStore", {
  state: () => {
    return {
      mlActive: false,
      calls: 0,
      spinnerConfig: spinnerConfig,
      spinnerEnums: spinnerEnums,
    };
  },
  getters: {
    mlTasksConfig() {
      return this.spinnerConfig?.tasksConfig || null;
    },
  },
  actions: {
    show() {
      this.calls++;
      this.mlActive = true;
    },
    hide() {
      this.calls--;
      if (this.calls == 0) {
        this.mlActive = false;
      }
      if (this.calls == 0) this._clearTasks();
    },
    configureModeSpinner() {
      this.spinnerConfig.mode = this.spinnerEnums.ModesSpinner.progress;
    },
    addTask(name, text) {
      this.spinnerConfig.tasksConfig.tasks.push({
        id: this.spinnerConfig.tasksConfig.tasks.length + 1,
        name,
        text,
        status: this.spinnerEnums.statusTask.pending,
      });
    },
    updateTask(name, status = null, text = null) {
      const indexTask = this.spinnerConfig.tasksConfig.tasks.findIndex(
        (task) => task.name === name,
      );
      if (status) {
        this.spinnerConfig.tasksConfig.tasks[indexTask].status = status;
      }
      if (text) {
        this.spinnerConfig.tasksConfig.tasks[indexTask].text = text;
      }
    },
    _clearTasks() {
      this.spinnerConfig.tasksConfig.tasks = [];
    },
  },
});
