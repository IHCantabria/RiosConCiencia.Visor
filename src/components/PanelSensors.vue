<template>
  <div class="panel-sensors">
    <vs-table
      ref="tableSensors"
      :data="sensors"
      :max-items="tableElements"
      :max-height="tableMaxHeight"
      pagination
      search
    >
      <template slot="thead">
        <vs-th sort-key="subTypeSensors">Sensor SubType</vs-th>
        <vs-th sort-key="typeSensor">Sensor Type</vs-th>
        <vs-th sort-key="name">Sensor</vs-th>
        <vs-th colspan="2" sort-key="selected">
          <span>Selected</span>
          <div @click="toggleAll">
            <vs-checkbox
              v-model="all"
              class="master-checkbox-sensors"
              icon="location_searching"
              :color="primaryColor"
            ></vs-checkbox>
          </div>
        </vs-th>
      </template>
      <template slot-scope="{ data }">
        <vs-tr v-for="(tr, indextr) in data" :key="indextr" :data="tr">
          <vs-td :data="tr.subTypeSensors">
            {{ tr.subTypeSensors }}
          </vs-td>
          <vs-td :data="tr.typeSensor">
            {{ tr.typeSensor }}
          </vs-td>
          <vs-td :data="tr.name">
            {{ tr.name }}
          </vs-td>
          <vs-td>
            <i
              class="fa fa-search"
              @click.stop="
                goToCoords({
                  type: tr.type,
                  coords: { lat: tr.latitude, lon: tr.longitude }
                })
              "
            />
          </vs-td>
          <vs-td :data="tr.selected">
            <div @click.stop="">
              <vs-checkbox
                v-model="tr.selected"
                icon="location_searching"
                :color="primaryColor"
              ></vs-checkbox>
            </div>
          </vs-td>
          <template slot="expand">
            <info-buoy
              v-if="tr.typeSensor == BUOYS.typeSensor"
              :open-license="tr.licenseOpen"
              :license-type="tr.license"
              :latitude="tr.latitude"
              :longitude="tr.longitude"
              :variables="tr.variables"
            ></info-buoy>
            <info-station
              v-else
              :open-license="tr.licenseOpen"
              :license-type="tr.license"
              :latitude="tr.latitude"
              :longitude="tr.longitude"
              :height="tr.height"
              :proyect-name="tr.project"
              :location="tr.placeName"
              :variables="tr.variables"
            ></info-station>
          </template>
        </vs-tr>
      </template>
    </vs-table>
  </div>
</template>
<script>
import { mapState, mapGetters, mapActions } from "vuex";
import { BUOYS } from "@/config/data-config";
import * as types from "@/store/types";
export default {
  components: {
    "info-buoy": () => import("@/components/ui/DataCardBuoy"),
    "info-station": () => import("@/components/ui/DataCardStation")
  },
  props: {
    tableElements: {
      type: Number,
      required: false,
      default: () => {
        return 15;
      }
    },
    tableMaxHeight: {
      type: String,
      required: false,
      default: () => {
        return "200px";
      }
    }
  },
  data() {
    return {
      all: false,
      BUOYS
    };
  },
  computed: {
    ...mapState({
      searchCoords: state => state.coords.searchCoords
    }),
    ...mapGetters({
      sensors: [types.G_GET_SENSORS_SORTED]
    })
  },
  watch: {
    sensors: {
      handler(newValue, oldValue) {
        if (oldValue.length == 0 && newValue.length != 0) this.initSensors();
      },
      deep: true
    }
  },
  methods: {
    ...mapActions({
      goToCoords: types.A_SET_GO_TO_COORDS
    }),
    initSensors() {
      this.all = false;
      this.setOrderDefault();
      this.clearExpandedBox();
      this.$nextTick(() => {
        !this.searchCoords.global ? this.selectDefault() : "";
      });
    },
    setOrderDefault() {
      this.$refs.tableSensors.currentSortKey = "distance";
      this.$refs.tableSensors.currentSortType = "asc";
    },
    selectDefault() {
      this.toggleAll();
      this.all = true;
    },
    toggleAll() {
      this.sensors.forEach(data => {
        data.selected = this.all ? false : true;
      });
    },
    clearExpandedBox() {
      while (document.getElementsByClassName("tr-expand").length > 0) {
        const element = document.getElementsByClassName("tr-expand")[0];
        element.remove();
      }
    }
  }
};
</script>
<style scoped lang="scss">
.panel-sensors {
  width: 100%;
}

.master-checkbox-sensors {
  margin-left: 10px;
}
</style>
