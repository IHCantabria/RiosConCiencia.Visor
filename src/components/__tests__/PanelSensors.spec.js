import { shallowMount, createLocalVue } from "@vue/test-utils";
import PanelSensors from "../PanelSensors.vue";
import merge from "lodash.merge";
import Vuex from "vuex";
import Vuesax from "vuesax";
import * as types from "@/store/types";

const localVue = createLocalVue();
localVue.use(Vuesax);
localVue.use(Vuex);

function createWrapper(overrides) {
  const defaultMountingOptions = {
    propsData: {},
    localVue
  };
  return shallowMount(PanelSensors, merge(defaultMountingOptions, overrides));
}

describe("panelSensors", () => {
  let store;
  beforeEach(() => {
    store = new Vuex.Store({
      modules: {
        coords: {
          state: {
            searchCoords: {}
          },
          actions: {
            [types.A_SET_GO_TO_COORDS]: jest.fn()
          }
        },
        apiSensors: {
          getters: {
            [types.G_GET_SENSORS_SORTED]: () => [
              {
                id: 51,
                mapID: "51B",
                selected: true,
                name: "VigoTG",
                type: "marker",
                typeSensor: "Buoy",
                subTypeSensors: "Mareógrafo",
                distance: 0,
                latitude: 42.24300003,
                longitude: -8.72599983,
                color: "blue"
              },
              {
                id: 52,
                mapID: "52B",
                selected: true,
                name: "VillagarciaTG",
                type: "marker",
                typeSensor: "Buoy",
                subTypeSensors: "Mareógrafo",
                distance: 0,
                latitude: 42.60100174,
                longitude: -8.77000046,
                color: "blue"
              },
              {
                id: 54,
                mapID: "54B",
                selected: true,
                name: "MarinTG",
                type: "marker",
                typeSensor: "Buoy",
                subTypeSensors: "Mareógrafo",
                distance: 0,
                latitude: 42.406101226,
                longitude: -8.6911001205,
                color: "blue"
              }
            ]
          }
        }
      }
    });
  });
  //UNIT TESTS
  it("call setOrderDefault when initSensors is called", () => {
    expect.assertions(1);
    const setOrderDefault = jest.fn();
    const wrapper = createWrapper({
      methods: {
        setOrderDefault
      },
      store
    });
    wrapper.vm.initSensors();
    expect(setOrderDefault).toHaveBeenCalled();
  });
  it("call clearExpandedBox when initSensors is called", () => {
    expect.assertions(1);
    const clearExpandedBox = jest.fn();
    const wrapper = createWrapper({
      methods: {
        clearExpandedBox
      },
      store
    });
    wrapper.vm.initSensors();
    expect(clearExpandedBox).toHaveBeenCalled();
  });
  it("call selectDefault when initSensors is called and searchCoords.global is false", async () => {
    expect.assertions(1);
    const selectDefault = jest.fn();
    const wrapper = createWrapper({
      methods: {
        selectDefault
      },
      store
    });
    wrapper.vm.initSensors();
    await wrapper.vm.$nextTick();
    expect(selectDefault).toHaveBeenCalled();
  });
  it("call toggleAll when selectDefault is called and changed data 'all'", () => {
    expect.assertions(2);
    const toggleAll = jest.fn();
    const wrapper = createWrapper({
      methods: {
        toggleAll
      },
      store
    });
    wrapper.vm.selectDefault();
    expect(toggleAll).toHaveBeenCalled();
    expect(wrapper.vm.all).toBe(true);
  });
  //SNAPSHOT TESTS
  it("renders correctly", () => {
    expect.assertions(1);
    const wrapper = createWrapper({
      store
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});
