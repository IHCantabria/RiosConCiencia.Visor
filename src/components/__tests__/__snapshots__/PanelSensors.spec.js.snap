// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`panelSensors renders correctly 1`] = `
<div
  class="panel-sensors"
>
  <vs-table-stub
    color="primary"
    currentpage="1"
    data="[object Object],[object Object],[object Object]"
    descriptionitems=""
    maxheight="200px"
    maxitems="15"
    nodatatext="No data Available"
    pagination="true"
    search="true"
    total="0"
  >
    <template>
      <vs-th-stub
        sortkey="subTypeSensors"
      >
        Sensor SubType
      </vs-th-stub>
       
      <vs-th-stub
        sortkey="typeSensor"
      >
        Sensor Type
      </vs-th-stub>
       
      <vs-th-stub
        sortkey="name"
      >
        Sensor
      </vs-th-stub>
       
      <vs-th-stub
        colspan="2"
        sortkey="selected"
      >
        <span>
          Selected
        </span>
         
        <div>
          <vs-checkbox-stub
            class="master-checkbox-sensors"
            color="primary"
            icon="location_searching"
            iconpack="material-icons"
            size="default"
          />
        </div>
      </vs-th-stub>
    </template>
  </vs-table-stub>
</div>
`;
